src/Starlight/View.hs

```diff
@@ -1,37 +1,83 @@
+{-# LANGUAGE DisambiguateRecordFields #-}
 {-# LANGUAGE NamedFieldPuns #-}
 module Starlight.View
 ( View(..)
-, scaleToView
-, scaleToViewZoomed
 , aspectRatio
 , deviceSize
+, lengthToWindowPixels
+  -- * Transforms
+, ClipUnits(..)
+, Zoomed(..)
+, toContext
+, toWindow
+, toZoomed
+, toSystem
+, transformToWindow
+, transformToZoomed
+, transformToSystem
+  -- * Viewport
+, clipTo
+  -- * Re-exports
+, module Geometry.Transform
 ) where
 
-import Control.Lens ((.~), (^.))
-import Data.Function ((&))
-import Linear.Matrix
-import Linear.V2
-import Linear.Vector
+import Control.Effect.Lift
+import Control.Lens ((&), (.~), (^.))
+import Data.Functor.Interval
+import Geometry.Transform
+import GL.Viewport
+import Linear.Exts
+import UI.Context as Context
+import UI.Window as Window
 import Unit.Length
 
 data View = View
-  { scale :: Int
-  , size  :: V2 Int
-  , zoom  :: Float
-  , focus :: V2 (Kilo Metres Float)
+  { ratio :: Int    -- ^ Ratio of window pixels per context pixel.
+  , size  :: V2 (Window.Pixels Int)
+  , zoom  :: Double
+  , scale :: Double
+  , focus :: V2 (Mega Metres Double)
   }
 
--- | Return a matrix transforming the [[-1,1], [-1,1]] interval to device coordinates.
-scaleToView :: (Applicative v, Traversable v, R2 v) => View -> v (v Float)
-scaleToView View{ scale, size } = scaled (pure 1 & _xy .~ 1 / (fromIntegral <$> size) ^* fromIntegral scale)
-
--- | Return a matrix transforming the [[-1,1], [-1,1]] interval to zoomed device coordinates.
-scaleToViewZoomed :: (Additive v, Applicative v, Traversable v, R2 v) => View -> v (v Float)
-scaleToViewZoomed vs@View{ zoom } = scaleToView vs !*! scaled (pure 1 & _xy .~ pure (1 / zoom))
-
 aspectRatio :: View -> Float
-aspectRatio View{ size } = size' ^. _x / size' ^. _y where
+aspectRatio View{ size } = size'^._x / size'^._y where
   size' = fromIntegral <$> size
 
-deviceSize :: View -> V2 Int
-deviceSize View{ scale, size } = scale *^ size
+deviceSize :: View -> V2 (Context.Pixels Int)
+deviceSize View{ ratio, size } = fmap Context.Pixels (ratio *^ fmap Window.getPixels size)
+
+lengthToWindowPixels :: View -> Double
+lengthToWindowPixels View{ zoom, scale } = 1/zoom * scale
+
+
+newtype ClipUnits a = ClipUnits { getClipUnits :: a }
+newtype Zoomed a = Zoomed { getZoomed :: a }
+
+
+toContext :: View -> Transform Double ClipUnits Context.Pixels
+toContext View{ size } = mkScale (pure 1 & _xy .~ 1 / (fromIntegral <$> size))
+
+toWindow :: View -> Transform Double Context.Pixels Window.Pixels
+toWindow View{ ratio } = mkScale (pure 1 & _xy .~ fromIntegral ratio)
+
+toZoomed :: View -> Transform Double Window.Pixels Zoomed
+toZoomed View{ zoom } = mkScale (pure 1 & _xy .~ pure (1/zoom))
+
+toSystem :: View -> Transform Double Zoomed (Mega Metres)
+toSystem View{ scale, focus } = mkScale (pure scale) >>> mkTranslation (ext (prj <$> negated focus) 0)
```

Comment:
Scale is now rightly a property of the presentation, and could conceivably be combined into the zoom at some point (tho this might make making zoomed or zoom-sensitive UI elements more difficult).